src/main/java/appeng/me/storage/ItemCellInventory.java

```diff
@@ -0,0 +1,316 @@
+			}
+		}
+		catch( final Throwable err )
+		{
+			return true;
+		}
+
+		return false;
+	}
+
+	public static boolean isCell( final ItemStack i )
+	{
+		if( i == null )
+		{
+			return false;
+		}
+
+		final Item type = i.getItem();
+		if( type instanceof IStorageCell )
+		{
+			if ( ( (IStorageCell) type ).getChannel() == AEApi.instance().storage().getStorageChannel( IItemStorageChannel.class ) )
+			{
+				return ( (IStorageCell) type ).isStorageCell( i );
+			}
+		}
+
+		return false;
+	}
+
+	public static void addBasicBlackList( final int itemID, final int meta )
+	{
+		BLACK_LIST.add( ( meta << Platform.DEF_OFFSET ) | itemID );
+	}
+
+	private static boolean isBlackListed( final IAEItemStack input )
+	{
+		if( BLACK_LIST.contains( ( OreDictionary.WILDCARD_VALUE << Platform.DEF_OFFSET ) | Item.getIdFromItem( input.getItem() ) ) )
+		{
+			return true;
+		}
+		return BLACK_LIST.contains( ( input.getItemDamage() << Platform.DEF_OFFSET ) | Item.getIdFromItem( input.getItem() ) );
+	}
+
+	@Override
+	public IAEItemStack injectItems( final IAEItemStack input, final Actionable mode, final IActionSource src )
+	{
+		if( input == null )
+		{
+			return null;
+		}
+		if( input.getStackSize() == 0 )
+		{
+			return null;
+		}
+
+		if( isBlackListed( input ) || this.cellType.isBlackListed( this.i, input ) )
+		{
+			return input;
+		}
+
+		final ItemStack sharedItemStack = input.createItemStack();
+
+		if( ItemCellInventory.isStorageCell( sharedItemStack ) )
+		{
+			final IMEInventory meInventory = getCell( sharedItemStack, null );
+			if( meInventory != null && !this.isEmpty( meInventory ) )
+			{
+				return input;
+			}
+		}
+
+		final IAEItemStack l = this.getCellItems().findPrecise( input );
+		if( l != null )
+		{
+			final long remainingItemSlots = this.getRemainingItemCount();
+			if( remainingItemSlots < 0 )
+			{
+				return input;
+			}
+
+			if( input.getStackSize() > remainingItemSlots )
+			{
+				final IAEItemStack r = input.copy();
+				r.setStackSize( r.getStackSize() - remainingItemSlots );
+				if( mode == Actionable.MODULATE )
+				{
+					l.setStackSize( l.getStackSize() + remainingItemSlots );
+					this.updateItemCount( remainingItemSlots );
+					this.saveChanges();
+				}
+				return r;
+			}
+			else
+			{
+				if( mode == Actionable.MODULATE )
+				{
+					l.setStackSize( l.getStackSize() + input.getStackSize() );
+					this.updateItemCount( input.getStackSize() );
+					this.saveChanges();
+				}
+				return null;
+			}
+		}
+
+		if( this.canHoldNewItem() ) // room for new type, and for at least one item!
+		{
+			final int remainingItemCount = (int) this.getRemainingItemCount() - this.getBytesPerType() * 8;
+			if( remainingItemCount > 0 )
+			{
+				if( input.getStackSize() > remainingItemCount )
+				{
+					final ItemStack toReturn = sharedItemStack.copy();
+					toReturn.setCount( sharedItemStack.getCount() - remainingItemCount );
+					if( mode == Actionable.MODULATE )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)